### PR TITLE
basic diff to detect differences between service definitions

### PR DESCRIPTION
**Pull request by zyndiecate**
This implements basic diff functionality to detect differences between service definitions. This will be necessary to implement service definition updates.

### CODE REVIEW

File: diff.go [hunk 1]
```diff
@@ -2,154 +2,119 @@ package userconfig
...
+	diffInfos = append(diffInfos, diffComponentRemoved(newDef.Components, oldDef.Components)...)
 
-	// Diff is called by diffHierarchy when two nodes have the same name and should be checked
-	// for internal differences. Diffs should be written to changes.
-	Diff(path []string, other node, changes chan<- DiffInfo)
+	return diffInfos
 }
 
-type appNode AppDefinition
+func diffServiceNameUpdated(a, b AppName) []DiffInfo {
```
Comment (ewoutp) [inline diff.go]: I don't like `a` and `b` as argument names. Please change to `old` and `new` or something like that.

**Comment by ewoutp** [comment]
The whole file deserves a few comments.

File: diff.go (hunk 1 above)
Comment (zyndiecate) [inline diff.go]: Sure. I was not happy with that either. Just wrote it down. Done.

File: diff.go [hunk 2]
```diff
@@ -1,155 +1,416 @@
...
-	// What type changed: app, service, component
 	Type DiffType
 
-	// Path to the node that changed.
-	Name []string
+	Old string
+	New string
+}
+
+func (di DiffInfo) Action() string {
```
Comment (ewoutp) [inline diff.go]: Please add comment describing the meaning and expected use of the return values.

File: diff.go [hunk 3]
```diff
@@ -1,155 +1,416 @@
...
+	case DiffInfoComponentRemoved:
+		return "remove component"
+	case DiffInfoComponentUpdated:
+		return "update component"
+	default:
+		panic("unknown diff type")
+	}
+}
+
+func (di DiffInfo) Reason() string {
```
Comment (ewoutp) [inline diff.go]: Please add comment describing the meaning and expected use of the return values.

File: diff.go [hunk 4]
```diff
@@ -1,155 +1,416 @@
...
+func diffComponentPorts(oldDef, newDef ComponentDefinition) []DiffInfo {
+	diffInfos := []DiffInfo{}
 
-func (n appNode) Name() string { return n.AppName }
-func (n appNode) Children() []node {
-	nodes := []node{}
-	for _, service := range n.Services {
-		nodes = append(nodes, serviceNode(service))
+	oldPorts := oldDef.Ports.String()
+	newPorts := newDef.Ports.String()
```
Comment (ewoutp) [inline diff.go]: Are you sure the order of ports is always the same?
If I change `80/tcp, 81/tcp` to `81/tcp, 80/tcp` I would not expect to get a change.

Comment (ewoutp) [inline diff.go]: Never mind, read the `String()` implementation.

File: diff_test.go [hunk 5]
```diff
@@ -28,151 +29,355 @@ func ExampleDefinition() AppDefinition {
...
+	}
 }
 
-func TestDiffComponentMultipleChanges(t *testing.T) {
-	oldCfg := ExampleDefinition()
-	newCfg := ExampleDefinition()
+// Because of our smart data types some definitions can be represented
+// differently. The following test ensures that a diff only is created in case
+// the component really changed.
+func TestDiffComponentDefinitionNoUpdate(t *testing.T) {
```
Comment (ewoutp) [inline diff_test.go]: Why is this function called `...NoUpdate` while it does expect diffs?

File: env.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
+package userconfig
```
Comment (ewoutp) [inline env.go]: Good that this is now in its own file

**Comment by ewoutp** [comment]
It took me a while to notice that the `String()` methods of the lists were all sorting their items.
Would be helpful if this behavior was documents in the comments on those functions.

File: expose.go [hunk 7]
```diff
@@ -10,6 +14,21 @@ type ExposeDefinition struct {
...
 }
 
+func (ed *ExposeDefinition) String() string {
+	m := map[string]string{
+		"port":        ed.Port.String(),
+		"component":   ed.Component.String(),
+		"target_port": ed.TargetPort.String(),
+	}
+
+	raw, err := json.Marshal(m)
```
Comment (ewoutp) [inline expose.go]: Will this always return the same order of map keys? 
I doubt it.

File: links.go [hunk 8]
```diff
@@ -19,6 +23,22 @@ type LinkDefinition struct {
...
 
+func (ld LinkDefinition) String() string {
+	m := map[string]string{
+		"service":     ld.Service.String(),
+		"component":   ld.Component.String(),
+		"alias":       ld.Alias,
+		"target_port": ld.TargetPort.String(),
+	}
+
+	raw, err := json.Marshal(m)
```
Comment (ewoutp) [inline links.go]: Will this always return the same order of map keys?

File: expose.go (hunk 7 above)
Comment (zyndiecate) [inline expose.go]: It does. Looks like the JSON implementation takes care about that: http://play.golang.org/p/U8nDgdga2X

Comment (ewoutp) [inline expose.go]: ok cool, anyway the unit tests should cover this if the go implementation changes at some point

File: diff.go (hunk 2 above)
Comment (zyndiecate) [inline diff.go]: done

File: diff.go (hunk 3 above)
Comment (zyndiecate) [inline diff.go]: done

File: diff_test.go (hunk 5 above)
Comment (zyndiecate) [inline diff_test.go]: I don't like the name either. Did not find a better one though. It is to ensure that a component that is not updated actually will not be updated. That is, not recognized as being different.

**Comment by zyndiecate** [comment]
> It took me a while to notice that the String() methods of the lists were all sorting their items.
Would be helpful if this behavior was documents in the comments on those functions.

Added comments.

File: expose.go (hunk 7 above)
Comment (zyndiecate) [inline expose.go]: Added tests.

File: links.go (hunk 8 above)
Comment (zyndiecate) [inline links.go]: Yes. See https://github.com/giantswarm/user-config/pull/60#discussion-diff-39385281

File: env.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	"github.com/juju/errgo"
+)
+
+// TODO this should be named EnvDefinition
```
Comment (ewoutp) [inline env.go]: Let's create an issue for that

File: links.go [hunk 10]
```diff
@@ -1,6 +1,11 @@
+// TODO this file should be called link.go
```
Comment (ewoutp) [inline links.go]: Let's create an issue for that

File: diff_test.go [hunk 11]
```diff
@@ -28,151 +29,355 @@ func ExampleDefinition() AppDefinition {
...
+			Type: DiffInfoComponentAdded,
+			New:  "redis2",
+		},
 	}
 
-	testDiffCallWith(t, newCfg, oldCfg, expectedDiffItems)
+	testDiffCallWith(t, oldDef, newDef, expectedDiffInfos)
 }
+
+// TODO test each definition difference inside a component definition
```
Comment (ewoutp) [inline diff_test.go]: Is this still applicable?

**Comment by ewoutp** [comment]
Minor issue creation left, otherwise WFN & LGTM

File: env.go (hunk 9 above)
Comment (zyndiecate) [inline env.go]: See https://github.com/giantswarm/user-config/issues/61

File: diff_test.go (hunk 11 above)
Comment (zyndiecate) [inline diff_test.go]: See https://github.com/giantswarm/user-config/issues/61

File: env.go (hunk 9 above)
Comment (zyndiecate) [inline env.go]: See https://github.com/giantswarm/user-config/issues/62

File: links.go (hunk 10 above)
Comment (zyndiecate) [inline links.go]: See https://github.com/giantswarm/user-config/issues/63

**Pull request merged**